Add is_obstructed?

smlovin2 opened the pull request:
- Adds FactoryGirl gem and as well as a factories folder for /spec and factory files for piece, player and game
- Moves rpsec gem into development and test group along with FactoryGirl
- Adds a model folder to /spec and created a piece_spec.rb file in that folder
- Adds a bunch of tests for the is_obstructed? method in piece model
- Modifies spec/rails_helper.rb so that rspec will automatically migrate before testing if necessary
- Adds .rspec file with --color so that rspec output will be colored which I feel makes it a little more user friendly :smile: 
- Adds the is_obstructed? method to piece model to detect if a piece's movement is obstructed when moving between two points
- is_obstructed? returns true if there is piece **between** the starting and ending space and false otherwise

**Notes and questions:**
- I'm not sure if this it the best solution, seem like it could still be cleaned up some, may be a little overly complicated
- There are still a few issues with rubocop I'm not sure what to do about:

> app/models/piece.rb:12:7: C: Rename is_obstructed? to obstructed?.
>   def is_obstructed?(row, column)
- It wants to remove "is" from all the methods like is_obstructed? is that fine or do we want it to be is_obstructed for sure?

> app/models/piece.rb:17:5 Use a guard clause instead of wrapping the code inside a conditional expression.
>     elsif moving_diagonally?(row, column)
- I'm not sure what to do with this one really

> spec/factories/game_factory.rb:1:1: C: Carriage return character detected.
> FactoryGirl.define do
- I'm also confused on how to fix this one as well, can't figure out how to get rid of the carriage return  character

smlovin2 committed:
Added factory_girl gem to Gemfile and added folders and files to spec

* Added factory girl to the Gemfile, while also moving rspec and factory girl into a development and test group
* Created a folder for model specs and started a piece_spec.rb file
* Created a folder for factories in /specs and added a factory for piece

smlovin2 committed:
Added game and player factories, added tests for is_obstructed?, modfied spec/rails_helper to auto migrate and added .rspec to get color for rspec output

* Needed to add player and game factories due to the associations in piece factory
* Added a hand full of tests to test for the different cases horizontal, vertical, diagonal and invalid input
* Added a couple of lines to spec/rails_helper.rb so that rspec will automatically migrate before testing if neccessary
* Noticed that rspec didn't have colored output which I feel makes it a little harder to understand so I added .rspec with --color so now the output is colored

smlovin2 committed:
Got a "rough draft" of is_obstructed? implemented, added a few more tests and updated piece factory

I Got an  is_obstructed? method put togther that does successfully pass tests but seems like it could probably be cleaned up a bit. I did separate the different cases into different methods as suggested but I'm not sure if the way I went through to check if the spaces in between were occupied was the best way or not. I had two ideas, either to loop through all the pieces and see if any of them were in the way or just check the spaces along the path and see if any of them are occupied. I eventually decided to do the latter of the two since it seemed like it made more sense only to check the spaces we care about instead of looking at every piece. This method worked pretty well, I think, for horizontal and vertical paths but for diaganol paths it was a little more complicated. So not sure if maybe the other way would've been better or if I just need to rethink of how to do my loops. Along with that, there also seem like there might be a DRY issue with the way it is now since there is a lot of similar code being used in multiple spots so might need see if there is away to fix that up as well. Other than that it seems to be working at the moment so we at least have something usable that can be cleaned up if needed.

I added some more tests to to test all the different directions (left, right, up, down, up and right, up and left, down and right, down and left) and I also added a test for when the piece does not move.

Found out that I needed to capitalize the q in queen for the type of the piece so the STI would work

smlovin2 committed:
Added check_path method to reduce redundant code

Tried to fix some of the DRY problems mentioned before by adding a check_path method that take a row and column direction as well as distance. It then check along the path in that direction for that distance looking to see if there are any pieces along the way. Still not sure this is the best solution but I think it is getting there. :smile:

smlovin2 committed:
Added some constants to make code a little clearer

smlovin2 committed:
Forgot to actually use the constants, also cleaned up some stuff for rubocop

mscoutermarsh commented on app/models/piece.rb:
```diff
@@ -8,4 +8,91 @@ class Piece < ActiveRecord::Base
+  LEFT = -1
+
+  def moving_horizontally?(row)
+    self.row == row
```
love these small methods.

I think you could be even more descriptive with the variable names.

``` Ruby
def moving_horizontally?(destination_row)
  row == destination_row
end
```

`self` would no longer be needed.

mscoutermarsh commented on app/models/piece.rb:
```diff
@@ -8,4 +8,91 @@ class Piece < ActiveRecord::Base
+    self.row > row && self.column < column
+  end
+
+  def is_horizontal_obstructed?(column)
```
I think that erroring on `is_`.

Is new to Rubocop.

So let's follow what rubocop says and remove `is_`.

I think `horizontally_obstructed?(destination_column)` works nicely

mscoutermarsh commented on app/models/piece.rb:
```diff
@@ -8,4 +8,91 @@ class Piece < ActiveRecord::Base
+    end
+  end
+
+  def check_path(row_dir, col_dir, dist)
```
Good to avoid any abbreviations.

`row_direction`,  `col_direction`, `distance` are totally fine

mscoutermarsh commented on app/models/piece.rb:
```diff
@@ -8,4 +8,91 @@ class Piece < ActiveRecord::Base
+    else
+      fail 'Input is invalid!'
+    end
+  end
```
another way to write this is with guard statements. it would look like this.

``` Ruby
  def is_obstructed?(row, column)
    return horizontal_obstructed?(row) if moving_horizontally?(row)
    return vertical_obstructed?(row) if moving_vertically?(column)
    return diagonal_obstructed?(row, column) if moving_diagonally?(row, column)

    fail 'Input is invalid!'
 end
```

mscoutermarsh commented on Gemfile:
```diff
@@ -32,13 +32,17 @@ group :doc do
+group :development, :test do
+  gem 'rspec-rails', '~> 3.0'
+  gem 'factory_girl_rails', '~> 4.0'
+end
```
:+1:

mscoutermarsh commented on spec/models/piece_spec.rb:
```diff
@@ -0,0 +1,124 @@
+require 'rails_helper'
+
+RSpec.describe Piece, type: :model do
+  describe 'is_obstructed?' do
```
For instance methods. Put a `#` infront of the method name in describe blocks.

example:
 `describe '#obstructed?'`

mo' info: http://betterspecs.org/#describe

mscoutermarsh commented on spec/models/piece_spec.rb:
```diff
@@ -0,0 +1,124 @@
+
+RSpec.describe Piece, type: :model do
+  describe 'is_obstructed?' do
+    it 'should return false for no horizontal obstructions moving to the right' do
```
I'll usually leave out `should`

`it 'returns false...'` works, less typing :)

mscoutermarsh commented on spec/models/piece_spec.rb:
```diff
@@ -0,0 +1,124 @@
+    end
+
+    it 'should return true for horizontal obstructions moving to the right' do
+      piece1 = FactoryGirl.create(:piece, row: 0, column: 3)
```
it's fine to use `piece` here.

mscoutermarsh commented on app/models/piece.rb:
```diff
@@ -8,4 +8,91 @@ class Piece < ActiveRecord::Base
+      cur_column += col_dir
+
+      obstructed |= game.pieces.find_by(row: cur_row, column: cur_column).present?
+    end
```
Another way to write this

``` Ruby
 def check_path(row_dir, col_dir, dist)
    current_row = row
    current_column = column

    (1...dist).each do
      current_row += row_dir
      current_column += col_dir

      return false if game.pieces.find_by(row: current_row, column: current_column).present?
    end

    true
  end
```

mscoutermarsh commented on app/models/piece.rb:
```diff
@@ -8,4 +8,91 @@ class Piece < ActiveRecord::Base
+    else
+      check_path(DOWN, NONE, self.row - row)
+    end
+  end
```
could make use of guard statement here

``` Ruby
return check_path(UP, NONE, row - self.row) if self.row < row

check_path(DOWN, NONE, self.row - row)
```

mscoutermarsh commented:
Looks great. Let me know if you have any questions.

I have NO IDEA whats up with that carriage return failure. Let me know if you still struggling with it, i'll pull down the code and give it a try.

smlovin2 committed:
Made made suggested changes from code review

- Added more descriptive variables, got rid of abbreviations and removed self
- Fixed issues with rubocop(removed "is_" from methods, added gaurd cases and fixed the carriage return failure)
- Added # to describe for obstructed? instance method
- Removed should from "it" statements in piece_spec.rb
- Changed piece1 to be just piece
- Refactored check_path using a return and if statement
